### ISSUE TITLE
Have "var" as syntactic sugar for explicit pointer refs/derefs

### ISSUE THREAD

**Issue by kvark** [gpuweb/gpuweb#1456]
This is a follow-up from the discussion in https://github.com/gpuweb/gpuweb/issues/1334#issuecomment-756432744 as well as #1455

Having real pointers with explicit "take reference" (operator `&`) and de-reference (operator `*`) is future-proof, and it opens the doors for use-cases where a local variable needs to be modified by a called function:
```rust
fn fun(inout_ptr: ptr<function, i32>) {
  *inout_ptr = 4;
}
```

However, we also don't want the noise of references/de-references to be required when working with function-local variables, such as loop iteration indices:
```rust
for (var i: i32 = 0; i<5; i = i + 1) {
  array[i] = 1.0;
}
```

I think we should have both. `var` would be syntactic sugar for a function-allocated piece of data that is de-referenced. We don't have to describe, at least at start, the actual way of declaring this piece of data ("decl_storage").
```rust
var x: f32;
x = 5.0; // works as l-value
const y: f32 = x; // works as r-value
```

When working with compound structures, the `.` operator takes precedence over de-referencing:
```rust
struct Foo {
  x: i32,
  y: f32,
};
fn bar(my_foo: ptr<function, Foo>) {
  *my_foo.x = 5;
  *my_foo.y = 1.0;
  const z: f32 = *my_foo.y; // r-value
}
```

When working with `var`, it works the same way, just with dereferencing hidden behind the sugar:
```rust
var foo: Foo; // function-local var
foo.x = 5; // is sugar for `*(&foo).x = 5;`
foo.y = 1.0; // is sugar for `*(&foo).y = 1.0;`
fun(&foo.x); // taking explicit pointer of a variable
```

**Comment by litherum** [comment]
I'm having trouble parsing this:

> When working with compound structures, the `.` operator takes precedence over de-referencing:
```
*my_foo.x = 5;
```

Does that mean it gets turned into `(*my_foo).x = 5` or `*(my_foo.x) = 5;`? From the description you gave "the `.` operator takes precedence over de-referencing" I'd expect the latter, but I don't see how that would typecheck, since `x` is not a pointer.

What am I missing?

**Comment by litherum** [comment]
```
fn fun(inout_ptr: ptr<function, i32>) {
  *inout_ptr = 4;
}
```

How would one call this function?

**Comment by litherum** [comment]
It seems that the only difference between this proposal and https://github.com/gpuweb/gpuweb/pull/1455 is that this one has `&` (parallel to `*`). Am I understanding correctly?

**Comment by litherum** [comment]
Meta comment: I think we have 4 proposals:
1. Do it like C++'s references (this is what we have in the spec now)
2. Do it like C's pointers (if I'm not mistaken, that's what this proposal is, and what https://github.com/gpuweb/gpuweb/pull/1455 would be if there was a `&` that was added to it)
3. Do it like GLSL / HLSL's (and Swift's?) `inout` keyword, which puts reference-semantics on the receiving variable instead of in the type: https://github.com/gpuweb/gpuweb/issues/1334
4. Do it like SPIR-V / LLVM, where there are no variables, there are only pointers to storage, and the pointers get names but the storage doesn’t get a name: I think @jdashg was describing this verbally [during today's F2F](https://docs.google.com/document/d/1Csyz-MF9u2iw7WmjENTCr4BUbxnTgPk9l0u7ux6DSL4/edit), but I don't think he wrote it down

First of all, if we're going to have `*`, we should have `&` for symmetry. Having one without the other seems like it would only cause problems for us down the road.

Given that, our first preference would be 2 because it's the most forward-looking. Hypothetical "real" pointers in the future would use the same type as our current "logical" pointers. Failing that, we would prefer 1, followed by 3. We really don't want 4, because that would lead authors to write things like

```
for (var i: i32 = 0; *i < 10; *i = *i + 1) { ... }
```

which would be super gross.

**Comment by litherum** [comment]
We'd also like to consider the possibility of restricting these pointer-y-things to just function parameters, for now. We think that would capture pretty much all of the use cases, and would help people avoid writing difficult-to-read code. (Yes, we understand that parameters are actually just variables, but we feel there's not much point to using these aliases without functions, and if you want to use references/pointers, you probably want to use a function anyway just to make your code cleaner.)

**Comment by kvark** [comment]
@litherum 
> Does that mean it gets turned into (*my_foo).x = 5 or *(my_foo.x) = 5;

It turns into `*(my_foo.x) = 5`. That's what I was trying to say by "the . operator takes precedence over de-referencing": first we take `.`, then we do `*`. Sorry for this being unclear!

> It seems that the only difference between this proposal and #1455 is that this one has & (as parallel to *). Am I understanding correctly?

That sounds about right from what I could gather from #1455. I like it in spirit but I'm a bit confused about the details, hence I wrote my vision here. For example,
```rust
var<in> in_x : i32;
  const x : ptr<in, i32> = in_x;
```
This means the variables are *actual* pointers, even though not defined as such. Contrary, in this proposal this would look like:
```rust
var<in> in_x : i32;
const x : ptr<in, i32> = &in_x;
```

> How would one call this function?

```rust
var x: i32;
fun(&x);
```

> Do it like C++'s references (this is what we have in the spec now)

From what I understand, right now we have something more complicated than C++ references, although I'm not even sure it matters? The status quo here needs to be re-adjusted anyway. In #1368 that is meant to describe what we have today, there is a magic thing happening where the same expression can be either a value or a pointer, depending on the context.

> Do it like C's pointers (if I'm not mistaken, that's what this proposal is,

Right.

> Do it like SPIR-V / LLVM, where there are no variables

I suggested this idea for consideration back in https://github.com/gpuweb/gpuweb/issues/1334#issuecomment-756179120
We can have both, however, which I think is a better path. It works similarly to the `loop/for` distinction, where one is the core API and the other one is built on top as sugar.

> We'd also like to consider the possibility of restricting these pointer-y-things to just function parameters, for now.

That sounds good to me! We can relax this later.
Overall, it sounds like we are in agreement.

**Comment by kainino0x** [comment]
> I'd expect the latter, but I don't see how that would typecheck, since `x` is not a pointer.

I had the same question, I assume we simply define the `thingofpointertype.x` to be what C would call `&(thingofpointertype->x)`.

**Comment by jdashg** [comment]
That's how #1455 defines it, yes. (this is effectively OpAccessChain)

**Comment by jdashg** [comment]
> Meta comment: I think we have 4 proposals:
> 
>     1. Do it like C++'s references (this is what we have in the spec now)
> 
>     2. Do it like C's pointers (if I'm not mistaken, that's what this proposal is, and what #1455 would be if there was a `&` that was added to it)
> 
>     3. Do it like GLSL / HLSL's (and Swift's?) `inout` keyword, which puts reference-semantics on the receiving variable instead of in the type: #1334
> 
>     4. Do it like SPIR-V / LLVM, where there are no variables, there are only pointers to storage, and the pointers get names but the storage doesn’t get a name: I think @jdashg was describing this verbally [during today's F2F](https://docs.google.com/document/d/1Csyz-MF9u2iw7WmjENTCr4BUbxnTgPk9l0u7ux6DSL4/edit), but I don't think he wrote it down
👍

> 
> First of all, if we're going to have `*`, we should have `&` for symmetry. Having one without the other seems like it would only cause problems for us down the road.
> 
> Given that, our first preference would be 2 because it's the most forward-looking. Hypothetical "real" pointers in the future would use the same type as our current "logical" pointers. Failing that, we would prefer 1, followed by 3. We really don't want 4, because that would lead authors to write things like
> 
> ```
> for (var i: i32 = 0; *i < 10; *i = *i + 1) { ... }
> ```
> 
> which would be super gross.

Well, for what it's worth, that's what it's doing under the hood. If we hated it we could have different `const i in 0..n` syntax, among other options. It's not a dead end because it's ugly with no other changes. (or we could never leave a local design maxima!)

I took a boids example and hand "optimized"-slash-made-explicit its var usage into ptrs and consts to see what's "actually" happening behind the syntactic sugar:
https://jdashg.github.io/misc/webgpu/wgsl-boids-ptr-compiled.html
(`const foo$n` is snapshot `n` of variable `foo`)

Lines like `*&cMass = *&cMass + pos;` in spir-v are effectively:
```spirv
$cMass = OpVariable($fvec2, Function)
[...]
$1 = OpLoad($cMass)
$2 = OpFAdd($1, $pos)
OpStore($cMass, $2)
```
`$cMass` is *already* a pointer, which is exactly how references are handled in e.g. c++. 

I think Myles had an interesting point earlier that choice/optimization/elision of implicit loads and stores is sort of the bread and butter of most compilers. I do feel that we are at a different balance point than e.g. GCC though with respect to how aggressively we want to invest in making mediocre code fast.

It's definitely unusual language-wise that `const` is 'better' than `var` for intermediates, at least before the compiler hopefully optimizes/elides vars *into* consts/intermediates. Making a pointer to a var excludes actual var elision, of course.

**Comment by dneto0** [comment]
Hey @kvark, in this example:

```rust
        var<in> in_x : i32;
        const x : ptr<in, i32> = &in_x; 
```

What is the type of `in_x` in the second statement?

**Comment by kvark** [comment]
@dneto0 I feel like there is a sub-question here, but I'm missing it.
For the purposes of type checker, the `in_x` has a type `i32`, just like its declaration states.

**Comment by kvark** [comment]
To follow up on @litherum confusion about the semantics of "." in this proposal, it's exactly the same as `OpAccessChain` in SPIR-V (and `*` is `OpLoad`).

**Comment by litherum** [comment]
I think we resolved today to go in this proposal's general direction, with 2 caveats:
1. The specification will use "reference" language do define what the meaning of a `var` variable
2. Make the `.` operator mean what it means in C, not SPIR-V

**Comment by dneto0** [comment]
> 2. Make the . operator mean what it means in C, not SPIR-V

FWIW, I didn't hear that as consensus.    I heard openness to exploring C's ->

**Comment by dneto0** [comment]
Copying from what I wrote on matrix.org:

I was thinking about the ref vs. pointer direction that @jdashg was going to write up. 
I think I see where you're headed and can see how to map it to a type system change. 

Basically, have two distinct types `ref<SC,T>` and `ptr<SC,T>`.  Then:
* Inside a function (for the sake of an example), `var x : i32;` declares lifetime-bounded distinct storage, and `x` has type `ref<function,i32>`.
* Assignment statement LHS = RHS; requires LHS to be `ref<*,T>` when the RHS is of type T. 
* address-of `&` takes a `ref<SC,T>` and produces `ptr<SC,T>` 
* Deref operator `*` takes a `ptr<SC,T>` and produces a `ref<SC,T>`.
* Also modify all the type rules that act on non-variable non-pointer values, so that things like arithmetic can work on ref types too.
    * E.g.    unary negation operator operates both on  i32 and on `ref<*,i32>`.  
     * For the latter we use the value in the referenced storage, and then act as if working on i32.

If I have this right, this morphs C's l-value r-value distinction into a distinction in the WGSL type system.  That would be fine with me.
 
With this direction, we have a policy choice at function boundaries. As a usability issue, we could ban ref-typed values as parameters and instead the use case would have ptr, like @kvark suggests above.

**Comment by kvark** [comment]
@dneto0 so this is clearly after MVP, and just to get a good sense of the motivation here, could you start by providing a good example of *where* we'd want to specify the `ref` type explicity? I.e. what problem does `ref` solve?

**Comment by dneto0** [comment]
After internal discussion, we also need the rule:

* You can't store a reference. 

Then examples:

```
struct S {
  m : i32;
};
var s : S;
fn f() {
   //  you can't store a reference, so the RHS here must be non-ref and non-pointer.
   // RHS must be i32, and so a load from s.m's storage is required to get the value to put into the storage for x.
   var x: i32 = s.m; 

   x = 3;   // Does not modify s.m.   Modifies contents of the storage for x.
}
```
Further:
* Can't do `var y : i32 = &s.m;`    This is ill-typed, because you can't store a pointer.
* Can't do `var p0 : ptr<private,i32> = ...` because pointer type is not storable.
* Can do `const p1 : ptr<private,S> = &s.m; ` This is well typed, and then later can do:  `*p1 = 5;` to modify s.m

**Comment by dneto0** [comment]
> could you start by providing a good example of where we'd want to specify the ref type explicity? I.e. what problem does ref solve?

I think it would be rare.  
* Can't store a ref, so no `var rvar : ref<...> =`
*  I suspect there is no appetite for allowing a `const r : ref<SC,T> = ...;`  That would hurt my brain. Let's not go there.
*  Maybe only as function parameter?   And even that is unpopular.

I was saying in the meeting that it's ok to invent the 'ref' type as a synthetic thing for the purpose of the spec and the compiler.  When getting lowering down to machine level, it melts away because we've translated code to a lower level of abstraction that requires fewer concepts.

Recall we agreed broadly that all expressions have known types (#723).  At the time I expressed a caveat about it because of the ptr/ref duality that was lurking underneath here if we wanted more regular-looking variable reads and updates.

Back to your earlier question: 

> @dneto0 I feel like there is a sub-question here, but I'm missing it.
For the purposes of type checker, the in_x has a type i32, just like its declaration states.

If every expression has a type, and `in_x` has type i32, then what does it mean to take the address of an i32 value?
If I can do `&in_x` then why can't I do `&5`.   Clearly I shouldn't, and something in the spec should tell me so.  In C/C++ we have l-value and r-value but WGSL does not.  Instead, it's the almost entirely the type system that tells you what operations apply.  That's why we want to say "you can do & on a ref" but not "you can do & on i32".

Things get a lot more murky when taking sub-references.  E.g. `a.b.c.d.e.f`.  A person and a compiler are doing a lot of unpacking to process that correctly.   At *each step* we should have a precise idea about what it is we're thinking about.  So that was the question underlying my "what is the type of in_x" question.  It can't be "i32", and we need something more carefully structured and specified.

Stepping way way back and into the ivory tower, there's a strong tradition ([denotational semantics](https://en.wikipedia.org/wiki/Denotational_semantics)) to have a guiding principle that:
- Every syntactically correct sub-expression should have a type
- So do statements and functions and every other syntactic object in the program.
- Perhaps those types are over weird domains.  (e.g especially functions, loops)
- Processing the semantics of a whole program can be broken down into: perform type checking (recursively) and processing semantics by tracking values in these weird domains.    E.g. the semantics of a loop is the fixedpoint of some value in a weird domain.

That's all very high powered stuff (and I don't think the WGSL grammar is quite right for that use), but I take a strong hint from that tradition that inventing a synthetic type like this is a wise thing to do so that we can write a tighter spec, generate better tests, and make better software.
(Ok, end sermon.)

**Comment by kvark** [comment]
@dneto0 thank you for elaboration in great detail, this is very helpful!

> I was saying in the meeting that it's ok to invent the 'ref' type as a synthetic thing for the purpose of the spec and the compiler. 

So far I'm not seeing any evidence that we'd need the `ref` to be available for users, so your suggestion sounds good.

As for the type checking of references, I don't think what you are describing is the only true way of approaching this.
The type system has its use and its limits, and its up to us to decide how much of it we need.

For example, Rust was designed towards rich type system, trying to encode a lot of constraints into it, and yet there isn't a "ref" type in a sense that you are suggesting. In Rust, you can still do `x = y` without any type system knowledge that `x` is being assigned to, other than the fact that it's mutable (which is orthogonal to being a reference). Rust references are exactly the `ptr<>` thing we are discussing here, they don't pretend to be the same thing as values.

The way you are suggesting us to go is to enforce the references as their own types, in addition to pointers and values. I'm not exactly sure how this helps the type system, since technically that would need to have implicit conversion from `ref<T>` to `T` (where `T` is expected). I think we also agreed previously to have no implicit casts, so this doesn't sound nice from the type system perspective.

The other way we could go is to say there are no references (at the type level). The type system would see an assignment as `i32 = i32` and be totally happy with it. But the validation logic would ensure that:
  - if there is a `&` then the operand must be a variable, or a part of a variable
  - if something is on the left of an assignment, then it has to be a variable, or a part of a variable.

That's it! I'd argue that the complexity of this is lower than trying to make the type system understand references.
And it's not like going with `ref` would let you avoid this validation logic: even if there is a `ref` on the left side, we *still* have to validate that it refers to a mutable location (i.e. writable storage buffer)...

> If I can do &in_x then why can't I do &5. Clearly I shouldn't, and something in the spec should tell me so. 

The type system may let you do it, but the validation will fail because `5` is not a variable.

Finally, #723 was discussed (at least in the way I see it) in the context type inference for right-hand-side expressions, so it's not about distinguishing ref from non-ref at all.

**Comment by kainino0x** [comment]
> Finally, #723 was discussed (at least in the way I see it) in the context type inference for right-hand-side expressions, so it's not about distinguishing ref from non-ref at all.

It's only a concern if values of type `ref<>` can exist (`ref<>` rvalue, or `const x: ref<>`, etc.) which I _\*think\*_ is a problem in Jeff's proposal (but not in yours).

**Comment by dneto0** [comment]
I disagree at the "what is good design" level.  Embedding the rules in the type system is the better way to do it IMHO.  It's a clean embedding.


> > If I can do &in_x then why can't I do &5. Clearly I shouldn't, and something in the spec should tell me so.

> The type system may let you do it, but the validation will fail because 5 is not a variable.

Let's generalize:

What tells me I can do   `&(a.b.c.d)`.    The `(a.b.c.d)` term can have its address taken, but it is very far from being a variable.

I do not want to build up an entirely different set of analyses that recurses over the AST when we already have a perfectly good system that already does that: the type system.

Your argument is just like saying "it doesn't feel like a type to me".  But we get to set our own definitions to serve our own purposes.


>  I'm not exactly sure how this helps the type system, since technically that would need to have implicit conversion from ref<T> to T (where T is expected)

That's the advantage of this ref/ptr thing over what I proposed in #1368.  There's no implicitness required if you write extra type rules.

For example, take unary logical negation:

Today:

<table>
<thead><th>Precondition<td>Conclusion
<tr><td>e : bool  <td>   ! e : bool 
</table>

Tomorrow:

<table>
<thead><th>Precondition<td>Conclusion
<tr><td>e : bool  <td>   ! e : bool <td> Logical negation. Yields true when e is false, and false when e is true.
<tr><td>e : ref&lt;S,bool&gt;<br>where S is a storage class<td> ! e : bool <td>Logical negation of a boolean in referenced storage.<br>  Yields `! evalue`, where <i>evalue</i> is the boolean value in the storage referenced by <i>e</i> at the time the expression is evaluated.<br> Does not modify the value in storage.
</table>

So it's two rules for the two cases you have to think about.

I was deliberately careful and wordy.  Editorially we'd make some definitions about what it means to "load" a reference value and end up saying something like this instead:

<table>
<thead><th>Precondition<td>Conclusion
<tr><td>e : bool  <td>   ! e : bool <td> Logical negation. Yields true when e is false, and false when e is true.
<tr><td>e : ref&lt;S,bool&gt;<br>where S is a storage class<td> ! e : bool <td>Logical negation of a boolean in referenced storage.<br>  Yields `! evalue`, where <i>evalue</i> is the boolean loaded from <i>e</i> at the time the expression is evaluated.
</table>

(E.g. "Loading a referenced value" means:
- reading all the memory locations used to represent any value of the "store type"
- and yielding the value of the store type corresponding to the values read from those locations.
- while accessing no other memory locations.
 These are specific requirements so that you can write a precise memory model.)

The point is: the type system is good machinery for doing what we need.  It's good to have more but simple rules than extra _kinds_ of rules on the side.

**Comment by kvark** [comment]
@dneto0 Looks like we are in a disagreement here about how far we should take the type system to get the the language that we want. Does it matter at this stage? The group has agreed to move forward with this issue's proposal, while adjusting the syntax for "." operator.
Are you suggesting to go back to the drawing board and try to introduce the `ref` into the type system for this? Or are you arguing about the need to have the `ref` in the specification right now, even if not exposed to the user?

If the latter, I'd like to see the `ref` suggestion as a follow-up, since it's a huge complication, and it's not strictly necessary here. Following up with this would allow us to see the clear trade-off that it proposes for the type system.

For example, to your question of "What tells me I can do &(a.b.c.d)", I wrote earlier that:
> if there is a & then the operand must be a variable, or a part of a variable

It may not be the most rigorous description of what's expected, but it's fairly intuitive, and we can start with it.

> I do not want to build up an entirely different set of analyses that recurses over the AST when we already have a perfectly good system that already does that: the type system.

I don't imply the implementations to do any more work than they do today. The type system we are describing for the spec is not directly linked to implementation complexity. So, in other words, neither do I want to build this different set of analyses, but it appears that it's not needed.

> So it's two rules for the two cases you have to think about.

Well, for negation being the unary operator it's two rules. For a ternary operator that's going to be 8 rules?

> The point is: the type system is good machinery for doing what we need. It's good to have more but simple rules than extra kinds of rules on the side.

Type system is great machinery! I don't think the "more but simple rules" applies to the `ref` proposal though. At least, it's not obvious to the point where we'd need to get `ref` right off the bat. We should see what it brings on the table.

**Comment by dneto0** [comment]
> The group has agreed to move forward with this issue's proposal, while adjusting the syntax for "." operator.

No, there isn't consensus.  I'll try to restate the options on the table.

>  It may not be the most rigorous description of what's expected, but it's fairly intuitive, and we can start with it. 

Sorry, I disagree. This is not precise enough.

**Comment by dneto0** [comment]
Option 1:  current state:  'ptr' type,  implicit decay to dereference as late as possible. Example:

          var a : i32: = 1;
          const ptra: ptr<function,i32> = a;   
          ptra = 2;  // updates contents of 'a' to store 2.

Cons: 
-  implicit dereference. 
- "Why can I assign a constant?"

Option 2:  like 1, but rename 'ptr' to 'ref'.   Motivation: pointers can only reference inside valid storage inside a variable:

          var a : i32: = 1;
          const ptra: ref<function,i32> = a;   
          ptra = 2;  // updates contents of 'a' to have the value 2.

Cons: 
- implicit dereference.
- "why can I assign to a constant?"
Pros: just like C++ references, but with different syntax.


Option 3: 
- instead of `const ... ref<...> = ...`, make 'ref' like its own declaration type inside a function body, along side 'const' and 'var'
- type of a var is the reference type.
- No 'ptr' at all.
- assignment statement: LHS must be reference type, RHS must be non-reference  (automatic decay)
- For expressions like a.b.c.d[12], prefer the "sub-reference" action to "dereference", to minimize load/store footprint.  
- function in/out parameters expressed as 'ref' formal parameter

          var a : i32: = 1;
          ref<function> aref: i32 = a; // Takes the reference to a's storage.
                                                         // When the 'var' can elide the storage class, then 'ref' can too.
          aref = 2;  // updates contents of 'a' to have the value 2.
          // Cannot re-point aref to some other storage.
          var b: i32 =50;
          aref = b; // Stores 50 into contents for variable 'a'.

         Also function in/out is expressed as 'ref' formal argument.

Pros:  C++ reference semantics.
Cons: No explicit "taking a reference" at call site.


Option 4:  Like 3, but also have distinct 'ptr' type.  
- address-of '&' operator converts reference to ptr
- unary-*  operator converts ptr to reference
- Policy choice: For in/out arguments, choose whether formal arguments can only be ptr (and hence require & at call site), or whether allow both ref and ptr.

Pros: 
- covers more of C++
- explicit & when passing a variable as reference parameter into a function
Cons: Covers more of C++.   Bigger language footprint to cover same functionality.

**Comment by dneto0** [comment]
Option 3 is essentially the proposal from #1334

**Comment by ben-clayton** [comment]
+1 for Option 3. It's very close to what we currently have, is easy to understand for people with C++ experience.

**Comment by dneto0** [comment]
Further to option 3:
*  You could take a ref-of-a-ref, e.g.:

              var nested : array<array<i32,5>,2>;
              ref row : array<i32,5> = nested[1];
              ref cell : i32 = row[2];   // RHS of ref-decl is must be a value of reference type. No decay. 
              cell = 12;   //  sets contents of nested[1][2] to value 12.

**Comment by litherum** [comment]
Is it possible to extend option 3 in the future to support hypothetical “real” pointers (“variable pointers”) in the future, without adding a new distinct type?

**Comment by alan-baker** [comment]
> Is it possible to extend option 3 in the future to support hypothetical “real” pointers (“variable pointers”), without adding a new distinct type?

We could eventually allow references to be reassigned. A possible syntax would be to a ref constructor usable on the RHS of an assignment statement. I'm not sure there's a great (and unambiguous) syntax beyond that.

**Comment by RobinMorisset** [comment]
I'd like to see an example of what you call option 4. I believe it is somewhat different from what I would prefer, which I'll call option 5.

In option 5, we have explicit pointers with & and *, like option 4, but no explicit "ref" keyword. If you want to take the adress of something, you do it explicitly:
```
const myref : ptr<i32, ...> = &a.b.c.d;
...
*myref = 42;
...
*myref = 43;
```

It would make it clear that these cannot be reassigned, it would make it clear where aliasing may happen, and it would be perfectly familiar to anyone who knows C. Naturally, we would still have `var`, so local variables would not need &/*. They could also be used for inout arguments: just pass a pointer. And once more it would be very explicit where we are modifying something coming from elsewhere versus a local variable.

**Comment by dneto0** [comment]
I apologize for misrepresenting the option.   I missed a clearly-important difference(!)

**Comment by dneto0** [comment]
Discussed in [2021-03-09 meeting](https://docs.google.com/document/d/1IQLS_nypwJLjaMZlejMSMC_vP7D2pm0EH1NtZ1TIvs0/edit)

**Comment by kvark** [comment]
Wanted to clarify explicitly, that @RobinMorisset 's (5) is the same as (4), which was just incorrectly described. I think we should do this 4/5 in some way. One of the points of confusion was the semantics of the "." operator:
  - (a) the body of this issue suggests that "." matches `OpAccessChain` in SPIR-V, so you'd write `pointer.x.y = 1.0`.
  - (b) follow the C++ semantics, which is what MSL does as well. So the very same statement would either look like `(*pointer).x.y = 1.0` or `pointer->x.y` if we also introduce the arrow operator.

We discussed this with @jdashg and agreed that (b) is a good compromise. Having pointers with (limited) C++ semantics is most familiar to anybody who will use pointers.

**Comment by dneto0** [comment]
> (a) the body of this issue suggests that "." matches OpAccessChain in SPIR-V, so you'd write pointer.x.y = 1.0.

For (a)
No, the LHS would have to be a reference, but a pointer is not a reference.
You would need the * to explicitly reference, and therefore you write:

        *(pointer.x.y) = 1.0;  // Still  is explicit store.

But because of precedence rules (same as in C), * and & are lower precedence than  .  and [i], so you could write this as:

        *pointer.x.y = 1.0;  

It's well specified that only the single scalar locations that are modified.

(b)  There is no need for another syntax.  See (a).
If you introduce -> syntax, then what it does is a combination of computing the pointer to a structure member, and then converting that pointer to a reference.   That's two jobs in one.

(See https://en.cppreference.com/w/cpp/language/operator_member_access#Built-in_member_access_operators  "The expression `E1->E2` is exactly equivalent to (*E1).E2 for built-in types; that is why the following rules address only `E1.E2`")

So these are the same on the LHS:
  
        *pointer.x.y = 1.0;
         pointer->x.y = 1.0;
         pointer.x->y = 1.0;

But they are not the same on the RHS:

         const x : f32 =  *pointer.x.y;  // Loads a single f32 from memory.
         const x2 : f32 = pointer->x.y;  // Loads a whole "x", and then extracts the scalar y from it.  So this is at least two words accessed
        const x3 : f32 = pointer.x->y;  // Loads only single f32 from memory.

Introducing -> confuses the issue about what memory locations are being accessed. It changes depending on whether it's being used as the LHS of an assignment, or elsewhere.  This is a recipe for confusion and accidental data races.

**Comment by dneto0** [comment]
> Loads a whole "x", and then extracts the scalar y from it.  So this is at least two words accessed

Hm.  Maybe that's overstated?  But I know I'd be confused about it.

We don't need the extra syntax.

**Comment by dneto0** [comment]
If you want option 4/5, then we suggest: 

*   Still keep 'var' for things at module scope (buffers, textures, workgroup, private), and at function scope ('var').
    *   Useful for saying ‘each var is distinct storage’, which helps expressing aliasing rules more easily.
        *   This becomes important for the general no-alias assumption in the Vulkan/SPIR-V programming model.
        *   Affects how we specify aliasing rules for WGSL in/out function formal parameters.
*   There are two types that refer to storage, and are duals of each other:
    *   ref&lt;SC,T>, the reference type of T in storage class SC
        *   Never written in WGSL source.  Used for type checking and analysis.
        *   var a: i32;  // 'a' has type ref&lt;function,i32>, _everywhere in its scope_
        *   Not storable.
    *   ptr&lt;SC,T>, the pointer type of T in storage class SC
        *   Written in WGSL source, as the type of
            *   a const
            *   a function formal parameter, to model in/out param
        *   Not storable (stays the same as currently spec’d)
*   Difference between ref and ptr is like C's lvalue and rvalue, but we'll model directly in the type system, rather than as "side data".
*   unary & is address-of:  converts a ref to corresponding ptr
*   unary \* is dereference: converts a ptr to corresponding ref
*   Unary & and unary \* precedence [similar to C](https://en.cppreference.com/w/c/language/operator_precedence):
    *   Higher than arithmetic
    *   Lower than struct-member-selection (.field), and array subscripting [i]
    *   Means C experience will translate over to WGSL:

            *p + *q;  // Sum of number stored at p and number stored at q

*   In assignment statement, LHS is a ref, RHS is not a ref
*   Member accessor (.field) , array accessor ([i]):
    *   maps ref to ref
    *   maps ptr to ptr
    *   Consequence for storing: you can read/write a nested value starting with a pointer, like this:

                                    struct Inner { b: i32; c: f32; }

                                    struct Outer { a: Inner; }

                                    var foo : Outer;

                                    const pstruct: ptr&lt;function,Outer> = &foo;

                                    // Stores 1 into locations for ‘b’ inside ‘foo’’ storage.

                                    // Because of precedence, compute pstruct.a.b first, to compute the

                                    // pointer to the locations for ‘b’ in ‘foo’.

                                    // Only accesses the 4 bytes for storing ‘b’.

			*pstruct.a.b = 1;    



*   Similarly, can read/write nested value starting with a reference, like this:

            // Same setup as before…


            // First compute ‘foo.a.b’ which is the ref&lt;function,i32> pointing at


            // the locations for ‘b’ in ‘foo’.


            // Then store 1 into the locations referenced by that ref.


            // Only accesses the 4 bytes for storing b.


            foo.a.b = 1;

*   For reading, there is still ‘delayed auto-decay’ from reference to value.
    *   Motivation:  _rules are designed so that, by spec, the fewest memory locations are accessed.  Tends to avoid races on variables shared across threads._
    *   If a reference-to-T is used where a T is required, you can convert the reference-to-T by doing an implicit dereference (load).
        *   But delay this as long as possible: Only do it if there is no other way to successfully type-check the program.
    *   Consider this, with same setup as before:
        *   const bcopy: i32 = foo.a.b;
    *   We have several paths to a correct type-check the RHS to match an i32.
        *   Strategy 1: delay ref-decay as long as possible:
            *   foo : ref&lt;function,Outer>
                *   foo.a : ref&lt;function,Inner>
                    *   foo.a.b: ref&lt;function,i32>
                        *   Load from that ref to produce i32.  4 bytes accessed
        *   Strategy 2: Other extreme: eagerly ref-decay.  Load as soon as possible
            *   foo : ref&lt;function,Outer>
                *   Load foo to produce transient foovalue : Outer.  8 bytes accessed
                    *   foo.a : Inner
                        *   foo.a.b: i32
    *   The “delayed auto-decay” rule says you should stay in the ‘ref’ domain as long as possible.  Only backtrack if necessary to correctly type-check the program.  So by spec, we don’t pursue strategy 2, because strategy 1 succeeded.
        *   This is just being honest about lvalue and rvalue, while modeling that in the type system.
        *   This is important to spec so we have a clear rule about which memory locations are accessed.  This is critical for making a clear and unambiguous memory model.
*   We rejected C’s arrow notation for now:
    *   Could say: if p is a pointer, then p->a is a reference to the ‘a’ in the struct pointed-to by p.
        *   Continuing example from above:

                pstruct->a.b = 1;     // Identical behaviour as *pstruct.a.b

*   No need for additional syntax of ->
*   C programmers already understand that array-indexing with pointers looks the same as regular variable.
*   In C:
    *   int r[10];
    *   int\* pr = r; // alternately &r[0]
    *   pr[3] = 1;   // Hmm. Why isn’t this non-explicitness a problem?
*   In WGSL:
    *   var r : array&lt;i32,10>;
    *   const pr : ptr&lt;function,array&lt;i32,10>> = &a;
    *   \*pr[3] = 1;  // write to the fourth element.

**Comment by RobinMorisset** [comment]
*pointer.x.y only works if we have a completely different definition of "." than C and C++. In both of these, it would be a type error to do pointer.x, as pointers are not structures (nor LValues a.k.a ref of structures)
I see this as a source of confusion for programmers. If we follow C/C++ operators instead, then your three examples:
```
*pointer.x.y
pointer->x.y
pointer.x->y
```
are not equivalent: the second is valid and does what you want, the other two are clearly invalid.
And this is true on both the LHS and the RHS.

We don't strictly need the -> operator, (*pointer).x.y would work as well. But I am concerned by this frankly super exotic definition of ".".

About the need to clearly define which locations are accessed, I fully agree. But this has no impact on pointers: *pointer is a reference, and then when you chain .x and .y, you'll stay in the reference world by your "delayed auto-decay" rule (not sure I would have specced it exactly in these words, but I fully agree with the idea). It is perfectly clear that you don't load/store the whole struct.

**Comment by dneto0** [comment]
What I'm describing is at least regular.  For example, & and * would be inverses of each other.

C/C++ pointers are too suggestive of things we don't want.  E.g.

        int g = 0;
        auto* pg = &g;
        pg[0] = 1;    //  Sets g to 1.    That's just weird, isn't it? 

If you want explicitness from pointer access, then don't follow C all the way.

**Comment by RobinMorisset** [comment]
I agree that this example is not something we want to allow. But from my point of view the problem is with C arrays, not general pointer syntax. I would just reject the example at pg[0] because pg is of type ptr<int> and not ref<array<..>> (ignoring the auto in this example).

**Comment by dneto0** [comment]
> I agree that this example is not something we want to allow. But from my point of view the problem is with C arrays, not general pointer syntax. I would just reject the example at pg[0] because pg is of type ptr and not ref<array<..>> (ignoring the auto in this example).

Fair.  But I had one more example to show:

    int g = 0;
    int* pg = &g;
    0[pg] = 1;   // does the same thing

YOLO.  :-)

**Comment by RobinMorisset** [comment]
First I'd like to say that I agree with Dzmitry: we are discussing three different points here:
1) What should the semantics of `.` and pointers be
2) Should we have a `->` operator (which would be syntactic sugar for `(*a).b`)
3) Should we have a traditional Lvalue vs Rvalue distinction in the spec, or express it equivalently through type rules and a spec-only `ref<...>` type.

I think we all agree that 1) is the most important, and should be decided first (and is independent from 3)).

About 1), Myles proposed some interesting example when we discussed this internally:
```
struct Foo {
    a : int32;
    b : int32;
};
struct Bar {
    c: int32;
    foo : Foo;
};
...
const d: ptr<Bar, private> = &something;
...
```
Then the following SPIRV:
```
%tmp1 = OpAccessChain %d %field_foo
%tmp2 = OpAccessChain %tmp1 %field_b
%tmp3 = OpLoad %tmp2
        OpStore %tmp3 %r
```
would correspond to `var r: int32 = (*d).foo.b` in my proposal and to `var r: int32 = *d.foo.b` in David's proposal (if I understood it correctly).

There is another interesting way of implementing this computation in SPIRV:
```
%tmp1 = OpLoad %d
%tmp2 = OpCompositeExtract %tmp1 %field_foo
%tmp3 = OpCompositeExtract %tmp2 %field b
        OpStore %tmp3 %r
```
Implementing this in my proposal would look like this:
```
const tmp: Bar = *d;
var r: i32 = tmp.foo.b;
```
I believe this would also work in David's proposal, but in his proposal it could also be written as `var r: i32 = (*d).foo.b`.

Computation number 2 is generally slower (as it loads more memory), and it can also create races that don't exist in computation number 1 (if for example another thread is accessing `(*d).foo.a`). So it seems good to me to make it so that C++ programmers using `(*d).foo.b` by habit get the fast code by default, and have to be make a temporary variable if they want to really load the whole structure.

Both of our proposal rely on the "Decayed delay rule". I'd like to present an alternative phrasing for it, that I believe to be equivalent and might be easier to follow for some:
- `.` and `[]` operators are overloaded, and in particular they behave like `OpCompositeExtract` on bare values (a.k.a rvalues), and like `OpAccessChain` on Lvalues (a.k.a. refs).
- The left-side of an assignment has to be an Lvalue (a.k.a. ref) and it creates an `OpStore`
- Every other use of an Lvalue (e.g. right side of an assignment, argument to an operator/function, condition of an if/while, etc..) converts that Lvalue to an Rvalue through inserting an `OpLoad`.

From there I think we get David's proposal by adding the following rules:
- `.` and `[]` are also overloaded to operate on pointers (behaving like `OpAccessChain`)
- `*` transforms a pointer to either a right-value (creating an `OpLoad`) or to an Lvalue (creating no SPIRV-code) depending on context.

Whereas we get my proposal by adding the following rules:
- `*` transforms a pointer to a Lvalue, which is a no-op in SPIRV terms.

In both of our proposals there is a difference with C++:
- In my proposal `p[42]` fails to compile if p is a pointer, requiring (*p)[42]. This can easily be diagnosed by the compiler with a helpful error message
- In David's proposal, `(*p).x` generates potentially inefficient code (although I expect LLVM to optimize it in most cases), and the programmers are expected to use the (more surprising to me) `*p.x` instead.

**Comment by litherum** [comment]
In the [above example](https://github.com/gpuweb/gpuweb/issues/1456#issuecomment-799774180):

&nbsp; | David's Proposal | Robin's Proposal
-|-|-
`*(d.foo.b)` (the same as `*d.foo.b`) | Computation 1 (fast code) | Compiler error
`(*d).foo.b` (the same as `d->foo.b`, if we have `->`) | Computation 2 (slow code) | Computation 1 (fast code)

**Comment by kvark** [comment]
@RobinMorisset excellent analysis, thank you!

> In David's proposal, (*p).x generates potentially inefficient code

What is appealing to me in D's (David's / original Dzmitry's) proposal is that the cost is explicit. Programmer would have to go out of their way to type `(*p).x` versus just `*p.x`. So there is a clear "opt-in" for the cost.

If I had to pick one of these proposals for WGSL, the deciding factor would be consistency with vars. Since vars already function as references, the `something.field` uses an `OpLoad` (or `OpStore` if on the left side of the assignment) implicitly, so the potential clarity of a `*` deref operator is already compromised in the language. Therefore, it's probably not worth it to try preserving the part of the deref purity for pointers only.

TL;DR: D's proposal has a benefit of being simple, explicit, and ergonomic. But it's not as consistent with `var`s, and it's not intuitive to a portion of developers who don't know SPIR-V. So we should probably look into adopting R's proposal.

**Comment by litherum** [comment]
> What is appealing to me in D's (David's / original Dzmitry's) proposal is that the cost is explicit.

I believe the cost is _more_ explicit in Robin’s proposal. You can still achieve the slow version in Robin’s proposal if you really want to; it just requires one extra line of code. In David’s proposal, you can get it just by rearranging where the parentheses lie.

const tmp: Bar = *d;
var r: i32 = tmp.foo.b;

**Comment by dneto0** [comment]
> Every other use of an Lvalue (e.g. right side of an assignment, argument to an operator/function, condition of an if/while, etc..) converts that Lvalue to an Rvalue through inserting an OpLoad.

There is also the address-of operator:  &
That converts an l-value back to a pointer.

I disagree with creating side-rules that can instead be folded into the type system, especially when those side rules are mostly invisible.

There are two reeasons for having  const pointers in the language at all:
- you want to compute an "address" early and reuse it many times elsewhere in the source
- support in/out params cleanly

The first use is almost always going to be a compound, e.g. "get a reference to field x of array element i in buffer foo".  That's a compound operation on a reference to a reference, or pointer to pointer.
It's a bad idea to inject a -> operator which only lets you do one step.

The argument seems "I like C, so give me C, except not really all of it because I don't like C arrays completely."

**Comment by dneto0** [comment]
I heard today @litherum claim that compounded sub-access indexing of pointer to pointer is uncommon or unusual, particularly in shader languages that humans write.

I'll grant that's true for MSL, given it inherits C++'s inability to do `.` on a pointer.

But it's actually visible in HLSL byte-addressible-buffers: you give a byte address, and 1,2,3 or 4 words of data to read/write.  So you can compute what is effectively an interior pointer to the buffer, through multiple nesting levels of structured-data by computing that byte address, as an integer.
That's a little low level to me because the data are always uints, but it's the "compounded pointer expression" is there in a very commonly used shader language.

Here's what I mean:

```
$ cat a.wgsl
struct Particle {
  [[offset(0)]] a: i32;
  [[offset(16)]] b: vec4<f32>;
};
[[block]] struct B {
  [[offset(0)]] data : [[stride(32)]] array<Particle,100>;
};
[[group(0), binding(0)]] var<storage> buf : [[access(read_write)]] B;

[[stage(compute)]]
fn main() -> void {
   buf.data[12].b.y = 1.0;
}

$ tint --format hlsl a.wgsl
struct Particle {
  int a;
  float4 b;
};
struct B {
  Particle data[100];
};

RWByteAddressBuffer buf : register(u0);

[numthreads(1, 1, 1)]
void main() {
  buf.Store((4 * 1) + 16 + (32 * 12) + 0, asuint(1.0f));  // Look at that address.  (o,o)
  return;
}

```
Now, look at that address calculation. That's ripe for strength reduction and common-subexpression elimination.  Adding a loop an indexing by the loop counter makes it more interesting:

```
$ cat b.wgsl
struct Particle {
  [[offset(0)]] a: i32;
  [[offset(16)]] b: vec4<f32>;
};
[[block]] struct B {
  [[offset(0)]] blah : vec4<f32>;
  [[offset(16)]] data : [[stride(32)]] array<Particle,100>;
};
[[group(0), binding(0)]] var<storage> buf : [[access(read_write)]] B;

[[stage(compute)]]
fn main() -> void {
   var i: i32 = 0;
   loop{
     if (i==100) {break;}
     buf.data[i].b.y = 1.0;
     i = i + 1;
   }
}

$ tint --format hlsl b.wgsl
struct Particle {
  int a;
  float4 b;
};
struct B {
  float4 blah;
  Particle data[100];
};

RWByteAddressBuffer buf : register(u0);

[numthreads(1, 1, 1)]
void main() {
  int i = 0;
  for(;;) {
    if ((i == 100)) {
      break;
    }
    buf.Store((4 * 1) + 16 + (32 * i) + 16, asuint(1.0f));
    i = (i + 1);
  }
  return;
}
```


The "compute an interior pointer" pattern appears repeatedly, though at a lower level:
* x86: [LEA vs. MOV](https://stackoverflow.com/questions/1658294/whats-the-purpose-of-the-lea-instruction)
* LLVM IR has getelementptr
* SPIR-V has AccessChain variants

**Comment by kvark** [comment]
To be fair, if you have structured data, you'll unlikely be using `RWByteAddressBuffer` in the first place. You'd instead write this as:
```hlsl
struct B0 {
    float4 blah;
};
RWStructuredBuffer<B0> buf0;
RWStructuredBuffer<float4> buf;
```

And you'd do the regular indexing in HLSL: `tb1[x] = float4(1.0, 0.0, 0.0, 1.0);`.
So the "uncommon or unusual, particularly in shader languages that humans write" claim still stands.

**Comment by dneto0** [comment]
Hey, I spoke abruptly at the last meeting saying there was no "performance" perfomance affect.  (Sorry for over-talking @litherum while he was trying to make his point.)

I think there is a misunderstanding at the heart of that disagreement.


That is, `*p` in the RHS region of an assignment, or in other expression evaluations is assumed (by the argument of @litherum and @RobinMorisset ) to always do the memory access (load) of the full footprint of what `p` is pointing to.

But this is not so, because it depends on an outer context.   For example, we could write.

        (*p).x

Which breaks down to:

         p   :  a  pointer to some complex  struture

        *p    : form a reference to that structure, in memory.    But don't yet access it.

        (*p)  : grouping, so still the same reference.


         (*p).x   :    form a memory reference to the 'x' member  in that structure

Then if we have that on the LHS, then we store to that.  If this expression is on the RHS and the non-pointer, non-reference value of it is demanded, then and only then do we load from that.

This is true both in C++, and in my proposal.

Repeating the example @litherum  gave in a table,

        (*d).foo.b 

No memory access is performed at all, until we have narrowed down the footprint to _only the size of the b member_.
Also, no memory access is performed unless the non-reference non-pointer value stored in the 'b' member in memory is demanded either by the surrounding assignment, or by a surrounding expression evaluation.

And because of precedence, in my proposal, with 'd' a pointer, we _can_ drop the parentheses and have exactly the same semantics


            *d.foo.b     // computes first d.foo.b as a pointer to the b thing in memory, then *(d.foo.b) a reference to that 'b' in memory.

**Comment by dneto0** [comment]
Discussed in meeting of [2021-03-23](https://docs.google.com/document/d/19inX49uQBMkWx1lSuPkFWJ8fGRJUGMNXmkUj5TJWnwY/edit#)

See update at https://github.com/gpuweb/gpuweb/issues/1530#issuecomment-805158676
